Added Floating Point arithmetic simplifications.

4y8 opened the pull request:
Added simplifications described in #10816.
```
x -. 0.0 = x
x *. 1.0 = x
x /. 1.0 = x
x *. 2.0 =  x +. x
x /. 2^N  =  x *. 2^{-N}
```
This optimization saves space in the generated code and avoids
unnecessary instructions. It shouldn't modify programs' behavior, as
this equalities work for all floating point numbers.

4y8 committed:
Added Floating Point arithmetic simplifications.

Added simplifications described in #10816.
```
x -. 0.0 = x
x *. 1.0 = x
x /. 1.0 = x
x *. 2.0 =  x +. x
x /. 2^N  =  x *. 2^{-N}
```
This optimization saves space in the generated code and avoids
unnecessary instructions. It shouldn't modify programs' behavior, as
this equalities work for all floating point numbers.

4y8 commented:
@xavierleroy In #10816, you mentioned that the last equality only works for appropriate values of N. I couldn't find which values are concerned as I am not familiar with CompCert's code base. Could you please tell me for which values equality works? Currently the equality is only applied to positive powers of 2.

4y8 committed:
Fixed typos.

The file was not saved before commiting. Bug fixes have therefore
not been taken into account by the first commit.

xavierleroy commented:
Everything I know about FP is distilled in this paper: https://xavierleroy.org/bibrefs/Boldo-Jourdan-Leroy-Melquiond-JAR.html

`x /. 2^N  =  x *. 2^{-N}` is valid if `|N| < 1023` or, equivalently, if both `2^N` and `2^{-N}` can be represented.

4y8 commented:
Thanks.
As the equality is currently applied, only to positive `N`, it should cause no problem, as 2^1023 and greater numbers can't be represented. I will read the paper to adjust the PR.

4y8 commented:
We would a way of knowing if a given floating number is a power of 2 and know which one. Currently, it only works when `N` is a strictly positive integer, as the program currently converts `2^N` to an integer (if it is one) and checks if it is equal to 2 power it log2.

xavierleroy commented:
> We would a way of knowing if a given floating number is a power of 2 and know which one.

Sure.  That's a nice little FP programming exercise.  Hint: have a look at `Float.frexp` and `Float.ldexp`.

4y8 commented:
Thanks. We can check that the float return by `Float.frexp` is 0.5.

4y8 committed:
Fixed division's optimization.

Made the division optimization rule work with all suitable values.

gasche commented:
There is some general context that I would like to understand first. We are talking of making the compiler codebase more complex by adding transformations, some of which appear non-trivial and easy to get wrong (I think that this is the second time @xavierleroy points out that one of the proposals was wrong). Plus reviewing/discussing the change takes time.

What are the benefits that correspond to these costs? Are we expecting some existing programs to become faster? (Which ones? Can we measure the performance improvement?) Does this unlock better ways to write programs? Why do we want to do this work?

4y8 committed:
Added a comment to clarify an optimisation

4y8 commented:
I also think that benchmarks are needed. I only implemented this changes because similar optimisations are made to integer arithmetics and because it was mentioned in #10816 that CompCert applied these simplifications.

xavierleroy commented:
You should also test for correctness, because right now I can see two bugs in your implementation of the division optimization.  I converted this PR to draft mode because it's not ready for review.

4y8 committed:
Fixed the division.

The ldexp expression was ill-formed.

4y8 commented:
I tested the division and it is now working. There is now a bug with multiplication because currently when calculating `x + x`, it currently calculates x twice (which is a waste of ressources and can cause problems with side effects). I don't see another way of doing it than transforming `2.0 *.x` to `let y = x in y + y`. But it generates an extra `let` construct. I will make it and check performances with benchmarks.

4y8 committed:
Added multiplication simplifications.

Now transform x *. 2.0 to let tmp = x in tmp +. tmp.

4y8 commented:
I made a small [benchmark](https://github.com/4y8/fp-benchmark-ocaml) and I saw an improvement of more than 10% in running time for multiplications by 2.0, and similar improvements for divisions by a power of two.

4y8 commented:
I found this [paper](http://perso.ens-lyon.fr/nicolas.brisebarre/Publi/fpdivision.pdf). I will implement algorithms 2 and 3, and check If the increase in performance is worth it, as this optimisation would add much more complexity than those currently implemented. Actually, they require having access to a hardware `fma`, which isn't always the case.

4y8 committed:
Removed unproved optimizations.

The optimization x -. 0.0, x + -0.0, x *. 1.0 and x /. 1.0 were
unproved and sometimes wrong so they were removed.

4y8 commented:
I removed all the unproved optimizations, and they all work. The benefit is of approximately 10%, on multiplication by 2 and division by a power of 2.

damiendoligez commented:
Something's wrong with the branch here. The PR appears to contain a lot of unrelated commits. Could you rebase onto trunk? Or if that doesn't work, reset your branch to trunk and then cherry-pick your own commits?

4y8 commented:
Yes, I just saw that. It's fixed now.

4y8 commented:
To sum up the PR, it implements the following optimisations, which were proved to be correct:
```
x *. 2.0 =  x +. x
x /. 2^N  =  x *. 2^{-N} (N < 1023)
```
Benchmarks show that it brings a speed-up of around 10% on the operations involved.

4y8 committed:
Merge branch 'trunk' into fp-arith-simplification

nojb commented on asmcomp/cmm_helpers.ml:
```diff
@@ -24,7 +24,7 @@ open Arch
 let bind name arg fn =
   match arg with
-    Cvar _ | Cconst_int _ | Cconst_natint _ | Cconst_symbol _ -> fn arg
+  | Cvar _ | Cconst_int _ | Cconst_natint _ | Cconst_symbol _ -> fn arg
```
This change (and other similar ones below) are unrelated to the PR.

nojb commented on asmcomp/cmm_helpers.ml:
```diff
@@ -217,16 +216,49 @@ let lsr_int c1 c2 dbg =
+  | c, Cconst_float(f, _) ->
+      (* x = 0.5 if and only if f is a power of 2 *)
+      let x, exp = Float.frexp f in
+      (* We can transform x/.2^{N} into x*.2^{-N} if and only if |N| < 1023,
```
Comment is misleading as `x` refers to `c` here.

nojb commented on asmcomp/cmm_helpers.ml:
```diff
@@ -217,16 +216,49 @@ let lsr_int c1 c2 dbg =
+      (* We can transform x/.2^{N} into x*.2^{-N} if and only if |N| < 1023,
+         because otherwise one of the may not be a valid floating point number
+         (2.**1023. evaluates to +infinity). *)
+      if x = 0.5 && abs exp < 1023
```
Is the bound correct? `f = x *. 2.0 ** exp`, so in your formula `N` is `exp - 1`, not `exp`.

4y8 committed:
Merge branch 'trunk' into fp-arith-simplification

4y8 committed:
Fixed the division by a power of two

The bound checked when making a division by a power of two was wrong by one.

4y8 commented:
The issues raised should be corrected.

silene commented on asmcomp/cmm_helpers.ml:
```diff
@@ -224,6 +223,38 @@ let asr_int c1 c2 dbg =
+      (* We can transform x/.2^{N} into x*.2^{-N} if and only if |N| < 1023,
+         because otherwise one of them may not be a valid floating point number
+         (2.**1023. evaluates to +infinity). Here, N=exp-1. *)
+      if x = 0.5 && abs exp - 1 < 1023
```
I cannot make sense of this test. Did you mean `abs (exp - 1) < 1023` instead?

Moreover, `2**1023` is not infinite (but `2**1024` is).

Finally, while we wrote `|N| < 1023` in the JAR paper, I wonder if it is a typo. It seems to me that `|N| <= 1023` would work just as well.

4y8 replied:
Sorry the `abs exp - 1` is a typo, because I originially wrote `abs exp`.

4y8 replied:
I made a commit to take into account this modification.

4y8 committed:
Fixed the bound in the `div_float` function

gasche commented:
This PR is in an awkward state. To put it frankly:

1. It is not clear that we care that much about the optimisations proposed. "10% faster in a benchmark" is not that exciting. (This being said, the benchmark is actually `Random.float (2. ** 65.) *. 2.`, whose cost are dominated by the `Random.float` call, so a real microbenchmark like `float i *. 2.` would give much stronger (artifical) numbers.)
2. On the other hand, the risk of making a mistake in the implementation and silently generating incorrect code in weird corner cases is rather high. This is clear from the fact that we have had this problem with other optimizations before, that iterations of this PR proposed several invalid optimizations, and that even some optimizations that are valid on paper have been proposed with implementation flaws. Finding out later that we are miscompiling floating-point code may be much more damaging to our users than the performance benefits of the optimization.
3. Given the two points above, the cold-blooded decision would be to say that we are not interested in those optimizations given the risk, and close the PR. (Or maybe: ask floating-point experts like @silene to rewrite them completely to get a strong confidence that the code is likely to be correct). But this would be harsh for the contributor @4y8 who made a nice effort on a first contribution on a difficult topic, so I would be sorry about that outcome.

@silene, what do you think we should do here?

silene commented:
I cannot comment on the style, but as far as the optimizations are concerned, they are definitely worth it. (No one should ever feel compelled to write `x *. 0.125`.) The thing I would have done differently is to write `ldexp 1.0 (-(exp - 1))` instead of `ldexp 2.0 (-exp)` (and then factor `let n = exp - 1 in` out). Also, I am a bit surprised that no tests are part of the pull request, but I don't know what kind of tests are part of the testsuite of the compiler, so this concern might be moot.

4y8 commented:
> But this would be harsh for the contributor @4y8 who made a nice effort on a first contribution on a difficult topic, so I would be sorry about that outcome.

I would definitely understand if you chose to do it.

> The thing I would have done differently is to write ldexp 1.0 (-(exp - 1)) instead of ldexp 2.0 (-exp) (and then factor let n = exp - 1 in out).

Your suggestion looks more readable ; I'll change the code.

> Also, I am a bit surprised that no tests are part of the pull request, but I don't know what kind of tests are part of the testsuite of the compiler, so this concern might be moot.

I can make a commit later this week including tests and the above suggestion, if you are okay with it.

4y8 committed:
Simplified the code and added tests.
